// @openzeppelin/upgrades-core/src/index.ts
export * from '@openzeppelin/upgrades-core/src/validate';
export { fetchOrDeploy, fetchOrDeployAdmin, fetchOrDeployGetDeployment } from '@openzeppelin/upgrades-core/src/impl-store';
export * from '@openzeppelin/upgrades-core/src/version';
export * from '@openzeppelin/upgrades-core/src/storage';
export {
  EIP1967BeaconNotFound,
  EIP1967ImplementationNotFound,
  getAdminAddress,
  getBeaconAddress,
  getImplementationAddress,
  toEip1967Hash,
  toFallbackEip1967Hash,
  isEmptySlot,
} from '@openzeppelin/upgrades-core/src/eip-1967';
export * from '@openzeppelin/upgrades-core/src/provider';
export * from '@openzeppelin/upgrades-core/src/src-decoder';
export * from '@openzeppelin/upgrades-core/src/solc-api';
export * from '@openzeppelin/upgrades-core/src/deployment';
export * from '@openzeppelin/upgrades-core/src/link-refs';
export * from '@openzeppelin/upgrades-core/src/error';

export {
  ManifestData,
  ImplDeployment,
  ProxyDeployment,
  Manifest,
  migrateManifest,
  DeploymentNotFound,
} from '@openzeppelin/upgrades-core/src/manifest';

export { getStorageLayoutForAddress } from '@openzeppelin/upgrades-core/src/manifest-storage-layout';

export * from '@openzeppelin/upgrades-core/src/scripts/migrate-oz-cli-project';

export { logWarning } from '@openzeppelin/upgrades-core/src/utils/log';
export { setProxyKind, processProxyKind, detectProxyKind } from '@openzeppelin/upgrades-core/src/proxy-kind';

export { UpgradeableContract } from '@openzeppelin/upgrades-core/src/standalone';

export { isTransparentOrUUPSProxy, isBeaconProxy, isTransparentProxy } from '@openzeppelin/upgrades-core/src/eip-1967-type';
export { getImplementationAddressFromBeacon, getImplementationAddressFromProxy } from '@openzeppelin/upgrades-core/src/impl-address';
export { isBeacon } from '@openzeppelin/upgrades-core/src/beacon';
export { addProxyToManifest } from '@openzeppelin/upgrades-core/src/add-proxy-to-manifest';

export {
  BeaconProxyUnsupportedError,
  LoadProxyUnsupportedError,
  PrepareUpgradeUnsupportedError,
  DeployBeaconProxyUnsupportedError,
  DeployBeaconProxyImplUnknownError,
  DeployBeaconProxyKindError,
  ForceImportUnsupportedError,
  NoContractImportError,
  ValidateUpdateRequiresKindError,
  PrepareUpgradeRequiresKindError,
  assertNotProxy,
} from '@openzeppelin/upgrades-core/src/usage-error';

export { ValidateUpgradeSafetyOptions, validateUpgradeSafety, ProjectReport, ReferenceContractNotFound } from '@openzeppelin/upgrades-core/src/cli';
